tool-openssl/req.cc

```diff
@@ -0,0 +1,675 @@
+
+  // Process each subject field
+  for (const auto& field : subject_fields) {
+    // Prompt with default value if available
+    if (field.default_value && field.default_value[0]) {
+      fprintf(stdout, "%s [%s]: ", field.short_desc, field.default_value);
+    } else {
+      fprintf(stdout, "%s: ", field.short_desc);
+    }
+    fflush(stdout);
+
+    // Get input with fgets
+    if (fgets(buffer, sizeof(buffer), stdin) == NULL) {
+      fprintf(stderr, "Error reading input\n");
+      return NULL;
+    }
+
+    // Remove newline character if present
+    size_t len = strnlen(buffer, sizeof(buffer));
+    if (len > 0 && buffer[len - 1] == '\n') {
+      buffer[len - 1] = '\0';
+      len--;
+    }
+
+    // Process input
+    const char* value = nullptr;
+    if (strncmp(buffer, ".", sizeof(buffer)) == 0) {
+      // Empty entry requested
+      value = "";
+    } else if (len == 0 && field.default_value) {
+      // Use default value
+      value = field.default_value;
+    } else if (len > 0) {
+      // Use provided input
+      value = buffer;
+    } else {
+      // Empty input and no default - use empty string
+      value = "";
+    }
+
+    // Only add non-empty values
+    if (value && value[0]) {
+      if (!X509_NAME_add_entry_by_NID(subj.get(), field.nid, chtype,
+                                     reinterpret_cast<const unsigned char*>(value),
+                                     -1, -1, 0)) {
+        fprintf(stderr, "Error adding %s to subject\n", field.field_name);
+        return NULL;
+      }
+    }
+  }
+
+  // If this is a CSR, handle extra attributes
+  if (isCSR) {
+    fprintf(stdout, "\nPlease enter the following 'extra' attributes\n");
+    fprintf(stdout, "to be sent with your certificate request\n");
+
+    // Process each extra attribute
+    for (const auto& attr : extra_attributes) {
+      // Prompt with default value if available
+      if (attr.default_value && attr.default_value[0]) {
+        fprintf(stdout, "%s [%s]: ", attr.short_desc, attr.default_value);
+      } else {
+        fprintf(stdout, "%s: ", attr.short_desc);
+      }
+      fflush(stdout);
+
+      // Get input with fgets
+      if (fgets(buffer, sizeof(buffer), stdin) == NULL) {
+        fprintf(stderr, "Error reading input\n");
+        return NULL;
+      }
+
+      // Remove newline character if present
+      size_t len = strnlen(buffer, sizeof(buffer));
+      if (len > 0 && buffer[len - 1] == '\n') {
+        buffer[len - 1] = '\0';
+        len--;
+      }
+
+      // Process input
+      const char* value = nullptr;
+      if (strncmp(buffer, ".", sizeof(buffer)) == 0) {
+        // Empty entry requested
+        value = "";
+      } else if (len == 0 && attr.default_value) {
+        // Use default value
+        value = attr.default_value;
+      } else if (len > 0) {
+        // Use provided input
+        value = buffer;
+      } else {
+        // Empty input and no default - use empty string
+        value = "";
+      }
```

Comment:
Except for the variable name (`attr` versus `field`), this logic is identical to what's above. We should split this out into its own function.